src/Type/Php/HashFunctionsReturnTypeExtension.php

```diff
@@ -0,0 +1,46 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\Constant\ConstantBooleanType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\MixedType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+use PHPStan\Type\TypeUtils;
+
+final class HashFunctionsReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'hash';
+	}
+
+	public function getTypeFromFunctionCall(FunctionReflection $functionReflection, FuncCall $functionCall, Scope $scope): Type
+	{
+		$defaultReturnType = ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+
+		if (!isset($functionCall->args[0])) {
+			return $defaultReturnType;
+		}
+
+		$argType = $scope->getType($functionCall->args[0]->value);
+		if ($argType instanceof MixedType) {
+			return TypeUtils::toBenevolentUnion($defaultReturnType);
+		}
+
+		$values = TypeUtils::getConstantStrings($argType);
+		if (count($values) !== 1) {
+			return TypeUtils::toBenevolentUnion($defaultReturnType);
+		}
+		$string = $values[0];
+
+		return in_array($string->getValue(), hash_algos(), true) ? new StringType() : new ConstantBooleanType(false);
```

Comment:
Would be useful/possible to directly emit a phpstan error when `hash()` is used with a hash-algo not defined in `hash_algos()`?

Reply:
You’ll get that error if you try to use false as a string.